### ISSUE TITLE
OAuth2.AcquireTokenWithClientCredentials returning token with incorrect credentials

### ISSUE THREAD

**Issue by KonstantynRomanowski** [microsoft/BCApps#1199]
**1. Describe the bug**
We are experiencing an issue with the OAuth2 authentication process. The problem seems to be related to the persistence of the Access token when using the OAuth2.AcquireTokenWithClientCredentials procedure.

We have a data structure that stores credentials for various web service connections, referred to as "endpoints." When we configure two endpoints with Microsoft Entra authentication and provide the following credentials:

**First Endpoint**: Incorrect Client Secret
**Second Endpoint**: Correct Client Secret

The following sequence of actions produces unexpected results:

- Test Connection with the First Endpoint: Results in HTTP 401 (Unauthorized) as expected due to incorrect credentials.
- Test Connection with the Second Endpoint: Test is successful, as expected.
- Test Connection with the First Endpoint again: Test is unexpectedly successful.

It seems that the Access token obtained during the second test is being reused for the first endpoint, even though the credentials for the first endpoint are incorrect. This suggests that the token is not being properly cleared or validated against the correct credentials.

**2. To Reproduce**

```
table 50100 "Endpoint"
{
    Caption = 'Endpoint';
    DrillDownPageID = "Endpoint List";
    LookupPageID = "Endpoint List";

    fields
    {
        field(1; "No."; Code[20])
        {
            DataClassification = CustomerContent;
            Caption = 'No.';
            NotBlank = true;
        }
        field(2; Description; Text[50])
        {
            DataClassification = CustomerContent;
            Caption = 'Description';
        }
        field(10; "Web Services Address"; Text[250])
        {
            DataClassification = CustomerContent;
            Caption = 'Web Services Address';
        }
        field(15; "Tenant ID"; Text[100])
        {
            DataClassification = CustomerContent;
            Caption = 'Tenant ID';
        }
        field(16; "Client ID"; Text[100])
        {
            DataClassification = CustomerContent;
            Caption = 'Client ID';
        }
        field(17; "Client Secret"; Text[100])
        {
            DataClassification = CustomerContent;
            ExtendedDatatype = Masked;
            Caption = 'Client Secret';
        }
    }
}
```

```
page 50100"Endpoint List"
{
    Caption = 'Endpoint List';
    CardPageID = "Endpoint Card";
    Editable = false;
    PageType = List;
    SourceTable = "Endpoint";
    UsageCategory = Administration;
    ApplicationArea = All;

    layout
    {
        area(content)
        {
            repeater(Group)
            {
                field("No."; Rec."No.")
                {
                    ApplicationArea = All;
                    ToolTip = 'Specifies the short endpoint number.';
                }
                field(Description; Rec.Description)
                {
                    ApplicationArea = All;
                    ToolTip = 'Specifies an endpoint description that will help you easily identify the endpoint.';
                }
            }
        }
    }
}
```

```
page 50101"Endpoint Card"
{
    Caption = 'Endpoint Card';
    PageType = Card;
    SourceTable = "Endpoint";
    
layout
    {
        area(content)
        {
            group(General)
            {
                Caption = 'General';
                field("No."; Rec."No.")
                {
                    ApplicationArea = All;
                }
                field(Description; Rec.Description)
                {
                    ApplicationArea = All;
                }
            }
            group("Web Service")
            {
                Caption = 'Web Service';
                field("Web Services Address"; Rec."Web Services Address")
                {
                    ApplicationArea = All;
                }
                group(AAD)
                {
                    ShowCaption = false;
                    field("Tenant ID"; Rec."AAD Tenant ID")
                    {
                        ApplicationArea = All;
                    }
                    field("Client ID"; Rec."AAD Client ID")
                    {
                        ApplicationArea = All;
                    }
                    field("Client Secret"; Rec."AAD Client Secret")
                    {
                        ApplicationArea = All;
                    }
                }
            }
        }
    }

    actions
    {
        area(processing)
        {
            action("Test Connection")
            {
                ApplicationArea = All;
                Image = TestFile;
                Caption = 'Test Connection';

                trigger OnAction()
                var TestConnection: Codeunit "Test Connection";
                begin
                    TestConnection.InvokeHttpRequest(Rec."No.");
                end;
          }
    }
}
```
```
codeunit 50100 "Test Connection"
{
local procedure GetAccessToken(Endpoint: Record "Endpoint") AccessToken: Text
    var
        OAuth2: Codeunit OAuth2;
        Scopes: List of [Text];
    begin
        Scopes.Add('https://api.businesscentral.dynamics.com/.default');
        Clear(OAuth2);
        if not OAuth2.AcquireTokenWithClientCredentials(Endpoint."Client ID", Endpoint."Client Secret", GetAuthorityUrl(Endpoint."Tenant ID"), '', Scopes, AccessToken) then
            Error('Failed to retrieve access token\ %1', GetLastErrorText());
    end;

    local procedure GetAuthorityUrl(TenantId: Text) Url: Text
    var
        AuthorityTxt: Label 'https://login.microsoftonline.com/{TenantId}/oauth2/v2.0/token', Locked = true;
    begin
        Url := AuthorityTxt;
        Url := Url.Replace('{TenantId}', TenantId);
    end;

    local procedure GetAuthenticationHeaderValue(AccessToken: Text) Value: Text;
    begin
        Value := StrSubstNo('Bearer %1', AccessToken);
    end;

procedure InvokeHttpRequest(Resource: Text; EndpointNo: Code[20]) Result: Text
    var
        RequestBlockedErr: Label 'Request was blocked by environment';
        RequestFailedErr: Label 'Request to connect to IC Partner failed. Error Code: %1  , Error Message: %2', Comment = '%1 = Error Code;  %2=Error Phrase';
        Client: HttpClient;
        Response: HttpResponseMessage;
        Url: Text;
        Error: Text;
        AccessToken: Text;
    begin
    
        Endpoint.Get(EndpointNo);
        AccessToken := GetAccessToken(ITIIICEndpoint);

        Client.DefaultRequestHeaders.Add('Authorization', GetAuthenticationHeaderValue(AccessToken));
        Client.DefaultRequestHeaders.Add('Accept', 'application/json');
        Url := Endpoint."Web Service Address; 
        if not Client.Get(Url, Response) then
            if Response.IsBlockedByEnvironment then
                Error(RequestBlockedErr)
            else
                Error(RequestFailedErr, GetLastErrorText());

        if not Response.IsSuccessStatusCode then 
            Error(RequestFailedErr, Response.HttpStatusCode, Response.ReasonPhrase);

        Response.Content.ReadAs(Result);
    end;
}
```

**3. Expected behavior**
The expected behavior is that the second attempt to test the connection with the first endpoint will fail as shown in the sequence below:

- Test Connection with the First Endpoint: Results in HTTP 401 (Unauthorized) as expected due to incorrect credentials.
- Test Connection with the Second Endpoint: Test is successful, as expected.
- Test Connection with the First Endpoint again: Results in HTTP 401 (Unauthorized) as expected due to incorrect credentials.

**4. Actual behavior**
The OAuth2.AcquireTokenWithClientCredentials procedure returns a valid token even when an incorrect Client Secret is provided, if a previous endpoint was tested with correct credentials.

**5. Versions:**

- AL Language. Pre-release Version : v13.0.921843 | Release Version: 12.7.964847
- Visual Studio Code: 1.77.3
- Business Central. Version: W1 24.0 (Platform 24.0.18621.0 + Application 24.0.16410.18679)
- List of Visual Studio Code extensions that you have installed: AL Language extension for Microsoft Dynamics 365 Business Central

### Final Checklist

Please remember to do the following:

* [ ] Search the issue repository to ensure you are reporting a new issue

* [ ] Reproduce the issue after disabling all extensions except the AL Language extension

* [ ] Simplify your code around the issue to better isolate the problem

**Comment by JesperSchulz** [comment]
@darjoo, could you triage this one, please?

**Comment by darjoo** [comment]
Hi @KonstantynRomanowski, I've communicated and opened an internal bug for our platform team. They will review and correct it.

Thank you for reporting this.
-Darrick

**Comment by peterelmqvist** [comment]
@darjoo can you confirm if this has been resolved and shipped in the product, if so in which version?

**Comment by JesperSchulz** [comment]
The bug is still active. This has not been resolved yet. @darjoo, could you follow up?

**Comment by darjoo** [comment]
@peterelmqvist It hasn't been resolved but is still an active bug as Jesper mentioned. The team that has the bug did acknowledge it and mentioned it is cached for a short period of time, however not a top priority fix afaik.
